plugins/gin/gin.go

```diff
@@ -0,0 +1,101 @@
+// Copyright 2019 Tetrate Labs
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/*
+Package http contains several client/server http plugin which can be used for integration with net/http.
+*/
+
+package gin
+
+import (
+	"fmt"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/gin-gonic/gin"
+	"github.com/tetratelabs/go2sky"
+	"github.com/tetratelabs/go2sky/propagation"
+	"github.com/tetratelabs/go2sky/reporter/grpc/common"
+)
+
+const (
+	httpServerComponentID int32 = 49
+)
+
+type routeInfo struct {
+	operationName string
+}
+
+type middleware struct {
+	routeMap     map[string]map[string]routeInfo
+	routeMapOnce sync.Once
+}
+
+//Middleware gin middleware return HandlerFunc  with tracing.
+func Middleware(engine *gin.Engine, tracer *go2sky.Tracer) gin.HandlerFunc {
+	if engine == nil || tracer == nil {
+		return nil
```

Comment:
We can't have this return nil as it will blow up the middleware chain. We can do two things:

1. change the function signature:
```go
func Middleware(engine *gin.Engine, tracer *go2sky.Tracer) (gin.HandlerFunc, error)
```

2. return an empty working Middleware:
```go
return func(c *gin.Context) {
    c.Next()
}
```

Reply:
Done.